In [19]:
import pandas as pd


In [20]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler  # for scaling data if needed
from sklearn.model_selection import train_test_split 

#For Regression Algorithms
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.neural_network import MLPRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import StandardScaler

#For Classification Algorithms
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

#Data Visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix
import os

In [21]:
# Loading the dataset
df = pd.read_csv('dataset.csv')


In [ ]:
# List of column names
list_of_column_names = list(df.columns)
print('List of column names: ', list_of_column_names)

# Selecting Features (Columns)

In [23]:
# Selecting columns to keep
columns_to_keep = [
    'Age_shift_and_scale',
    'Gender_shift_and_scale',
    'ECG_shift_and_scale',
    'CKMB_shift_and_scale',
    'TROP-I_shift_and_scale',
    'LAD_shift_and_scale',
    'LCA_shift_and_scale',
    'RCA_shift_and_scale',
    'Systolic_shift_and_scale',
    'Diastolic_shift_and_scale',
    'Cholesterol_shift_and_scale',
    'Chest_Pain_shift_and_scale',
    'Diabetic_shift_and_scale',
    'PHF /family history_shift_and_scale',
    'MI',
]

In [24]:
# Keeping only specified columns
df = df[columns_to_keep]

# Splitting the data into features and targets

In [25]:
# Splitting the data into features and target
X = df[['Age_shift_and_scale',
        'Gender_shift_and_scale',
        'ECG_shift_and_scale',
        'CKMB_shift_and_scale',
        'TROP-I_shift_and_scale',
        'LAD_shift_and_scale',
        'LCA_shift_and_scale',
        'RCA_shift_and_scale',
        'Systolic_shift_and_scale',
        'Diastolic_shift_and_scale',
        'Cholesterol_shift_and_scale',
        'Chest_Pain_shift_and_scale',
        'Diabetic_shift_and_scale',
        'PHF /family history_shift_and_scale']]
y = df['MI']  # Target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [26]:
# Feature Scaling (if necessary)
scaler = StandardScaler()
X_train_scaled_sampled = scaler.fit_transform(X_train)
X_test_scaled_sampled = scaler.transform(X_test)


#  Regression Models

Modelling

In [27]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

lin_reg = LinearRegression()
lin_reg.fit(X_train_scaled_sampled, y_train)

# Training R-squared
train_r2 = lin_reg.score(X_train_scaled_sampled, y_train)

# Testing R-squared
y_pred_test = lin_reg.predict(X_test_scaled_sampled)
test_r2 = r2_score(y_test, y_pred_test)

print("Linear Regression - Train R-squared:", train_r2)
print("Linear Regression - Test R-squared:", test_r2)



from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score

dt_reg = DecisionTreeRegressor()
dt_reg.fit(X_train, y_train)

# Training R-squared
train_r2 = dt_reg.score(X_train, y_train)

# Testing R-squared
y_pred_test = dt_reg.predict(X_test)
test_r2 = r2_score(y_test, y_pred_test)

print("Decision Tree Regression - Train R-squared:", train_r2)
print("Decision Tree Regression - Test R-squared:", test_r2)


from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score

rf_reg = RandomForestRegressor()
rf_reg.fit(X_train, y_train)

# Training R-squared
train_r2 = rf_reg.score(X_train, y_train)

# Testing R-squared
y_pred_test = rf_reg.predict(X_test)
test_r2 = r2_score(y_test, y_pred_test)

print("Random Forest Regression - Train R-squared:", train_r2)
print("Random Forest Regression - Test R-squared:", test_r2)


from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import r2_score

gbm_reg = GradientBoostingRegressor()
gbm_reg.fit(X_train, y_train)

# Training R-squared
train_r2 = gbm_reg.score(X_train, y_train)

# Testing R-squared
y_pred_test = gbm_reg.predict(X_test)
test_r2 = r2_score(y_test, y_pred_test)

print("Gradient Boosting Regression - Train R-squared:", train_r2)
print("Gradient Boosting Regression - Test R-squared:", test_r2)


from sklearn.svm import SVR
from sklearn.metrics import r2_score
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

svr_reg = SVR()
svr_reg.fit(X_train_scaled, y_train)

# Training R-squared
train_r2 = svr_reg.score(X_train_scaled, y_train)

# Testing R-squared
y_pred_test = svr_reg.predict(X_test_scaled)
test_r2 = r2_score(y_test, y_pred_test)

print("SVR Regression - Train R-squared:", train_r2)
print("SVR Regression - Test R-squared:", test_r2)


from sklearn.neural_network import MLPRegressor
from sklearn.metrics import r2_score
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

mlp_reg = MLPRegressor()
mlp_reg.fit(X_train_scaled, y_train)

# Training R-squared
train_r2 = mlp_reg.score(X_train_scaled, y_train)

# Testing R-squared
y_pred_test = mlp_reg.predict(X_test_scaled)
test_r2 = r2_score(y_test, y_pred_test)

print("Neural Network Regression - Train R-squared:", train_r2)
print("Neural Network Regression - Test R-squared:", test_r2)


from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import r2_score

knn_reg = KNeighborsRegressor()
knn_reg.fit(X_train, y_train)

# Training R-squared
train_r2 = knn_reg.score(X_train, y_train)

# Testing R-squared
y_pred_test = knn_reg.predict(X_test)
test_r2 = r2_score(y_test, y_pred_test)

print("K-Nearest Neighbors Regression - Train R-squared:", train_r2)
print("K-Nearest Neighbors Regression - Test R-squared:", test_r2)



Linear Regression - Train R-squared: 0.509282620440847
Linear Regression - Test R-squared: 0.5378423200348161
Decision Tree Regression - Train R-squared: 1.0
Decision Tree Regression - Test R-squared: 0.772113640942978
Random Forest Regression - Train R-squared: 0.9751874017042333
Random Forest Regression - Test R-squared: 0.8490366814426757
Gradient Boosting Regression - Train R-squared: 0.8384849705661939
Gradient Boosting Regression - Test R-squared: 0.7772596415465602
SVR Regression - Train R-squared: 0.8210519198486876
SVR Regression - Test R-squared: 0.8553690588988845
Neural Network Regression - Train R-squared: 0.8690382540210843
Neural Network Regression - Test R-squared: 0.8540276125385691
K-Nearest Neighbors Regression - Train R-squared: 0.860254692110315
K-Nearest Neighbors Regression - Test R-squared: 0.8605802941769092


c:\Users\devan\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


| Model                        | Train R-squared | Test R-squared |
|------------------------------|-----------------|----------------|
| Linear Regression            | 0.5093          | 0.5378         |
| Decision Tree Regression     | 1.0000          | 0.7371         |
| Random Forest Regression     | 0.9759          | 0.8450         |
| Gradient Boosting Regression | 0.8385          | 0.7751         |
| SVR Regression               | 0.8211          | 0.8554         |
| Neural Network Regression    | 0.8659          | 0.8487         |
| K-Nearest Neighbors Regression | 0.8603        | 0.8606         |


Data Visualization

In [28]:
output_dir_1 = "D:/btech-final-year-project/Data Augmentation/Outputs_of_Feature-Models/Regression"

In [29]:
# Linear Regression
plt.scatter(y_test, y_pred, color='blue')
plt.title('Linear Regression Model')
plt.xlabel('Actual Values')
plt.ylabel('Predicted Values')
plt.savefig(os.path.join(output_dir_1, "linear_regression_plot.png"))
plt.close()

# Decision Tree Regression
plt.scatter(y_test, y_pred, color='green')
plt.title('Decision Tree Regression Model')
plt.xlabel('Actual Values')
plt.ylabel('Predicted Values')
plt.savefig(os.path.join(output_dir_1, "decision_tree_regression_plot.png"))
plt.close()

# Random Forest Regression
plt.scatter(y_test, y_pred, color='red')
plt.title('Random Forest Regression Model')
plt.xlabel('Actual Values')
plt.ylabel('Predicted Values')
plt.savefig(os.path.join(output_dir_1, "random_forest_regression_plot.png"))
plt.close()

# Gradient Boosting Regression
plt.scatter(y_test, y_pred, color='orange')
plt.title('Gradient Boosting Regression Model')
plt.xlabel('Actual Values')
plt.ylabel('Predicted Values')
plt.savefig(os.path.join(output_dir_1, "gradient_boosting_regression_plot.png"))
plt.close()

# Support Vector Regression (SVR)
plt.scatter(y_test, y_pred, color='purple')
plt.title('SVR Model')
plt.xlabel('Actual Values')
plt.ylabel('Predicted Values')
plt.savefig(os.path.join(output_dir_1, "svr_regression_plot.png"))
plt.close()

# Neural Network Regression
plt.scatter(y_test, y_pred, color='brown')
plt.title('Neural Network Regression Model')
plt.xlabel('Actual Values')
plt.ylabel('Predicted Values')
plt.savefig(os.path.join(output_dir_1, "neural_network_regression_plot.png"))
plt.close()

# K-Nearest Neighbors Regression
plt.scatter(y_test, y_pred, color='black')
plt.title('K-Nearest Neighbors Regression Model')
plt.xlabel('Actual Values')
plt.ylabel('Predicted Values')
plt.savefig(os.path.join(output_dir_1, "knn_regression_plot.png"))
plt.close()

# Classification Models

Modelling

In [30]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, classification_report

# Create a pipeline with feature scaling and logistic regression
pipe_logreg = make_pipeline(StandardScaler(), LogisticRegression())

# Define a grid of hyperparameters to search over
param_grid_logreg = {
    'logisticregression__C': [0.001, 0.01, 0.1, 1.0, 10.0]  # Regularization parameter
}

# Perform grid search with cross-validation
grid_logreg = GridSearchCV(pipe_logreg, param_grid_logreg, cv=5, scoring='accuracy')
grid_logreg.fit(X_train, y_train)

# Get the best model and evaluate on train and test data
best_logreg = grid_logreg.best_estimator_
y_train_pred_logreg = best_logreg.predict(X_train)
y_test_pred_logreg = best_logreg.predict(X_test)

train_accuracy_logreg = accuracy_score(y_train, y_train_pred_logreg)
test_accuracy_logreg = accuracy_score(y_test, y_test_pred_logreg)

print("Best Parameters for Logistic Regression:", grid_logreg.best_params_)
print("Train Accuracy for Logistic Regression:", train_accuracy_logreg)
print("Test Accuracy for Logistic Regression:", test_accuracy_logreg)
print("Classification Report for Logistic Regression:\n", classification_report(y_test, y_test_pred_logreg))



from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report

# Train a decision tree classifier with max_depth parameter for pruning
dt_classifier = DecisionTreeClassifier(max_depth=5)  # Adjust max_depth as needed
dt_classifier.fit(X_train, y_train)

y_train_pred_dt = dt_classifier.predict(X_train)
y_test_pred_dt = dt_classifier.predict(X_test)

train_accuracy_dt = accuracy_score(y_train, y_train_pred_dt)
test_accuracy_dt = accuracy_score(y_test, y_test_pred_dt)

print("Train Accuracy for Decision Tree Classifier:", train_accuracy_dt)
print("Test Accuracy for Decision Tree Classifier:", test_accuracy_dt)
print("Classification Report for Decision Tree Classifier:\n", classification_report(y_test, y_test_pred_dt))




from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import accuracy_score, classification_report

# Define the parameter grid for random forest
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [5, 10, 20],
    'min_samples_split': [2, 5, 10]
}

# Perform randomized search with cross-validation
rf_classifier = RandomForestClassifier()
grid_rf = RandomizedSearchCV(rf_classifier, param_distributions=param_grid_rf, n_iter=10, cv=5, scoring='accuracy', random_state=42)
grid_rf.fit(X_train, y_train)

# Get the best model and evaluate on train and test data
best_rf = grid_rf.best_estimator_
y_train_pred_rf = best_rf.predict(X_train)
y_test_pred_rf = best_rf.predict(X_test)

train_accuracy_rf = accuracy_score(y_train, y_train_pred_rf)
test_accuracy_rf = accuracy_score(y_test, y_test_pred_rf)

print("Best Parameters for Random Forest Classifier:", grid_rf.best_params_)
print("Train Accuracy for Random Forest Classifier:", train_accuracy_rf)
print("Test Accuracy for Random Forest Classifier:", test_accuracy_rf)
print("Classification Report for Random Forest Classifier:\n", classification_report(y_test, y_test_pred_rf))




from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report

# Train an SVM classifier with regularization parameter C
svc_classifier = SVC(C=1.0)  # Adjust C value for regularization
svc_classifier.fit(X_train, y_train)

y_train_pred_svc = svc_classifier.predict(X_train)
y_test_pred_svc = svc_classifier.predict(X_test)

train_accuracy_svc = accuracy_score(y_train, y_train_pred_svc)
test_accuracy_svc = accuracy_score(y_test, y_test_pred_svc)

print("Train Accuracy for SVC:", train_accuracy_svc)
print("Test Accuracy for SVC:", test_accuracy_svc)
print("Classification Report for SVC:\n", classification_report(y_test, y_test_pred_svc))




from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report

# Train a k-NN classifier with scaled data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

knn_classifier = KNeighborsClassifier()
knn_classifier.fit(X_train_scaled, y_train)

y_train_pred_knn = knn_classifier.predict(X_train_scaled)
y_test_pred_knn = knn_classifier.predict(X_test_scaled)

train_accuracy_knn = accuracy_score(y_train, y_train_pred_knn)
test_accuracy_knn = accuracy_score(y_test, y_test_pred_knn)

print("Train Accuracy for K-NN Classifier:", train_accuracy_knn)
print("Test Accuracy for K-NN Classifier:", test_accuracy_knn)
print("Classification Report for K-NN Classifier:\n", classification_report(y_test, y_test_pred_knn))

Best Parameters for Logistic Regression: {'logisticregression__C': 10.0}
Train Accuracy for Logistic Regression: 0.9572895277207393
Test Accuracy for Logistic Regression: 0.9540229885057471
Classification Report for Logistic Regression:
               precision    recall  f1-score   support

           0       0.95      0.92      0.93       182
           1       0.98      1.00      0.99       266
           2       0.91      0.92      0.92       161

    accuracy                           0.95       609
   macro avg       0.95      0.95      0.95       609
weighted avg       0.95      0.95      0.95       609

Train Accuracy for Decision Tree Classifier: 0.9634496919917864
Test Accuracy for Decision Tree Classifier: 0.9573070607553367
Classification Report for Decision Tree Classifier:
               precision    recall  f1-score   support

           0       0.97      0.92      0.94       182
           1       0.99      0.98      0.99       266
           2       0.90      0.96     

| Model                     | Train Accuracy | Test Accuracy |
|---------------------------|----------------|---------------|
| Logistic Regression       | 0.9573         | 0.9540        |
| Decision Tree Classifier  | 0.9634         | 0.9573        |
| Random Forest Classifier  | 0.9856         | 0.9721        |
| Support Vector Classifier | 0.9630         | 0.9704        |
| K-Nearest Neighbors       | 0.9676         | 0.9704        |


Data Visualizations

In [37]:
output_dir_2 = "D:/btech-final-year-project/Data Augmentation/Outputs_of_Feature-Models/Classification"

In [38]:
import os
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix, classification_report

def plot_confusion_matrix(model_name, y_true, y_pred, output_dir):
    """
    Generate and save a confusion matrix plot for the specified model.
    
    Args:
        model_name (str): Name of the model for the plot title.
        y_true (array-like): True labels.
        y_pred (array-like): Predicted labels.
        output_dir (str): Directory to save the plot.
    """
    plt.figure(figsize=(8, 6))
    cm = confusion_matrix(y_true, y_pred)
    sns.heatmap(cm, annot=True, cmap='Blues', fmt='g')
    plt.title(f'Confusion Matrix - {model_name}')
    plt.xlabel('Predicted')
    plt.ylabel('Actual')
    plt.savefig(os.path.join(output_dir, f'confusion_matrix_{model_name.lower().replace(" ", "_")}.png'))
    plt.close()

# Visualize confusion matrices for each classification model
plot_confusion_matrix("Logistic Regression", y_test, y_test_pred_logreg, output_dir_2)
plot_confusion_matrix("Decision Tree Classifier", y_test, y_test_pred_dt, output_dir_2)
plot_confusion_matrix("Random Forest Classifier", y_test, y_test_pred_rf, output_dir_2)
plot_confusion_matrix("Support Vector Classifier", y_test, y_test_pred_svc, output_dir_2)
plot_confusion_matrix("K-Nearest Neighbors", y_test, y_test_pred_knn, output_dir_2)
